app/routers/menstrual_predictor.py

```diff
@@ -0,0 +1,179 @@
+    except SQLAlchemyError:
+        logger.info(
+            f'User {new_signed_up_to_period.user_id}'
+            'already signed up to the service, hurray')
+        db.rollback()
+    url = '/'
+    generate_predicted_period_dates(
+        db, data['avg_period_length'], last_period_date, current_user.id)
+
+    return RedirectResponse(url=url, status_code=HTTP_302_FOUND)
+
+
+@router.get("/get-period-dates/")
+def get_period_days(request: Request, db: Session = Depends(get_db)):
+    period_days = get_all_period_days(db, 1)
+    gaps_list = []
+    for i in range(len(period_days) - 1):
+        gap = get_date_diff(period_days[i].start, period_days[i + 1].start)
+        gaps_list.append(gap.days)
+    logger.critical(get_list_avg(gaps_list))
+
+
+def get_avg_period_gap(db: Session, user_id):
+    period_days = get_all_period_days(db, user_id)
+    gaps_list = []
+
+    for i in range(len(period_days) - 1):
+        gap = get_date_diff(period_days[i].start, period_days[i + 1].start)
+        gaps_list.append(gap.days)
+    return get_list_avg(gaps_list)
+
+
+def get_date_diff(date_1: datetime, date_2: datetime):
+    return date_2 - date_1
+
+
+def get_list_avg(received_list: List):
+    return sum(received_list) // len(received_list)
+
+
+def remove_existing_period_dates(db: Session, user_id: int):
+    (db.query(Event).
+     filter(Event.owner_id == user_id).
+     filter(Event.category_id == MENSTRUAL_PERIOD_CATEGORY_ID).
+     filter(Event.start > datetime.datetime.now()).
+     delete())
+    db.commit()
+    logger.info('Removed all period predictions to create new ones')
+
+
+def generate_predicted_period_dates(
+        db: Session,
+        period_length: str,
+        period_start_date: datetime,
+        user_id: int):
+    delta = datetime.timedelta(int(period_length))
+    period_end_date = period_start_date + delta
+    create_event(db, 'period', period_start_date, period_end_date,
+                 user_id, category_id=MENSTRUAL_PERIOD_CATEGORY_ID)
+    get_all_period_days(db, user_id=user_id)
+
+
+def add_3_month_predictions(db, period_length, period_start_date, user_id):
+    avg_gap = get_avg_period_gap(db, user_id)
+    avg_gap_delta = datetime.timedelta(avg_gap)
+    for i in range(4):
+        generate_predicted_period_dates(
+            db, period_length, period_start_date, user_id)
+        period_start_date += avg_gap_delta
+        logger.error(period_start_date)
+    logger.error(avg_gap_delta)
+
+
+def get_all_period_days(session: Session, user_id: int) -> List[Event]:
+    """Returns all period days filtered by user id."""
+
+    try:
+        period_days = sorted((session.query(Event).
+                              filter(Event.owner_id == user_id).
+                              filter(Event.category_id
+                                     == MENSTRUAL_PERIOD_CATEGORY_ID).
+                              all()), key=lambda d: d.start)
+
+    except SQLAlchemyError as err:
+        logger.debug(err)
```

Comment:
consider logger.error as this is error

Reply:
Or even better: `logger.exception` :)
(Which is technically `logger.error`, just with traceback)

Reply:
done